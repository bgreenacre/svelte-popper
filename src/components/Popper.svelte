<div {...props} bind:this={contentRef} class={classes} data-placement={placement}>
  {#if children}
    {children}
  {:else}
    <slot />
  {/if}
</div>
<script>
  import classnames from 'classnames';
  import PopperJS from 'popper.js';
  import { onDestroy, tick } from 'svelte';
  import exclude from './exclude.js';

  export let className = undefined;
  export let placement = 'bottom';
  export let eventsEnabled = true;
  export let positionFixed = false;
  export let modifiers = {};
  export let children = undefined;
  export let targetRef = undefined;
  export let arrowRef = undefined;
  export let outOfBoundaries = undefined;

  let classes;
  let contentRef;
  let popper = undefined;
  let props;
  let statePlacement = placement;

  const updateStateModifier = {
    enabled: true,
    order: 900,
    fn: (data) => {
      statePlacement = data.placement;
      outOfBoundaries = data.hide;

      return data;
    },
  };

  const getOptions = () => ({
    placement,
    eventsEnabled,
    positionFixed,
    modifiers: {
      ...modifiers,
      arrow: {
        ...(modifiers && modifiers.arrow),
        enabled: !! arrowRef,
        element: arrowRef,
      },
      updateStateModifier,
    },
  });

  const destroyPopperInstance = () => {
    if ( ! popper) return;

    popper.destroy();
    popper = undefined;
  };

  const updatePopperInstance = async () => {
    await tick();

    destroyPopperInstance();

    popper = new PopperJS(targetRef, contentRef, getOptions());
  };

  export const scheduleUpdate = () => popper && popper.scheduleUpdate();

  onDestroy(() => destroyPopperInstance());

  $: if (! targetRef) {
    throw new Error('A valid target reference must be passed to Popper component');
  }

  $: if (PopperJS.placements.indexOf(placement) === -1) {
    throw new Error(`Invalid placement sent: '${placement}' is not one of ${PopperJS.placements.join(', ')}.`);
  }

  $: props = exclude($$props, [
      'placement',
      'eventsEnabled',
      'positionFixed',
      'modifiers',
      'children',
      'targetRef',
      'arrowRef',
      'className',
      'popper',
    ]);

  $: classes = classnames(className, 'svlt-popper');
  $: arrowRef, targetRef, placement, positionFixed, modifiers, updatePopperInstance();
  $: if (popper) {
    if (eventsEnabled) {
      popper.enableEventListeners();
    }
    else {
      popper.disableEventListeners();
    }

    if (statePlacement) {
      scheduleUpdate();
    }
  }
</script>
