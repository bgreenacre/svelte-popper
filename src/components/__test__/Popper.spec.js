import Popper from '../Popper.svelte';
import { render, cleanup } from '@testing-library/svelte';

beforeEach(cleanup);

describe('Popper', () => {
  test('should render a popper with expected markup', () => {
    const targetRef = document.createElement('div');
    const { container } = render(Popper, { props: { targetRef }});

    expect(container).toMatchSnapshot();
  });

  test('should render with expected content', () => {
    const targetRef = document.createElement('div');
    const { container } = render(Popper, { props: { targetRef, children: 'Hello world!' }});
    const popper = container.querySelector('.svlt-popper');

    expect(popper.innerHTML.trim()).toBe('Hello world!');
  });

  test('should render with custom class name', () => {
    const targetRef = document.createElement('div');
    const { container } = render(Popper, { props: { targetRef, className: 'custom-class' }});
    const popper = container.querySelector('.svlt-popper');

    expect(popper.className).toContain('custom-class');
  });

  test('should throw exception when no targetRef given', () => {
    expect(() => {
      render(Popper);
    }).toThrow();
  });

  test('should throw exception when invalid placement given', () => {
    const targetRef = document.createElement('div');

    expect(() => {
      render(Popper, { props: { targetRef, placement: 'not valid' }});
    }).toThrow();
  });

  test('should destroy popper instance on unmount', () => {
    const targetRef = document.createElement('div');
    const wrapper = render(Popper, { props: { targetRef }});
    const { popper } = wrapper.component.$$.ctx;

    wrapper.unmount();
    expect(popper.state.isDestroyed).toBe(true);
  });
});
