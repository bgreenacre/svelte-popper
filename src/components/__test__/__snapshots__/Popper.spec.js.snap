// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popper should render a popper with expected markup 1`] = `
<body>
  <div>
    <div
      class="svlt-popper"
      data-placement="bottom"
    />
  </div>
</body>
`;
